File: android/src/main/java/com/supermap/smNative/SMCartography.java
package com.supermap.smNative;

import com.supermap.data.*;
import com.supermap.interfaces.mapping.SMap;
import com.supermap.mapping.*;

public class SMCartography {

    /**
     * 根据文本对齐中文返回文本对齐类型
     * @param alignmentText
     * @return
     */
    public static TextAlignment getTextAlignment(String alignmentText){
        switch (alignmentText) {
            case "TOPLEFT":
                return TextAlignment.TOPLEFT;
            case "TOPCENTER":
                return TextAlignment.TOPCENTER;
            case "TOPRIGHT":
                return TextAlignment.TOPRIGHT;
            case "BASELINELEFT":
                return TextAlignment.BASELINELEFT;
            case "BASELINECENTER":
                return TextAlignment.BASELINECENTER;
            case "BASELINERIGHT":
                return TextAlignment.BASELINERIGHT;
            case "BOTTOMLEFT":
                return TextAlignment.BOTTOMLEFT;
            case "BOTTOMCENTER":
                return TextAlignment.BOTTOMCENTER;
            case "BOTTOMRIGHT":
                return TextAlignment.BOTTOMRIGHT;
            case "MIDDLELEFT":
                return TextAlignment.MIDDLELEFT;
            case "MIDDLECENTER":
                return TextAlignment.MIDDLECENTER;
            case "MIDDLERIGHT":
                return TextAlignment.MIDDLERIGHT;
            default:
                return null;
        }
    }

    public static Geometry getGeoText(Recordset recordset) {
        try {
            if (recordset != null && recordset.getRecordCount() > 0) {
                recordset.moveFirst();
                Geometry geometry = recordset.getGeometry();
                if (geometry.getType() == GeometryType.GEOTEXT) {
                    return geometry;
                }
            }

            return null;
        } catch (Exception e) {
            e.printStackTrace();
            return null;
        }
    }

    public static Recordset getRecordset(int geometryID, String layerName) {
        try {
            Layers layers = SMap.getSMWorkspace().getMapControl().getMap().getLayers();
            Layer layer = layers.get(layerName);
            if (layer != null && layer.getDataset() instanceof DatasetVector) {
                DatasetVector datasetVector = (DatasetVector) layer.getDataset();
                int[] ids = {geometryID};
                return datasetVector.query(ids, CursorType.DYNAMIC);
            }
            return null;
        } catch (Exception e) {
            e.printStackTrace();
            return null;
        }
    }

    /**
     * 获取栅格图层设置类
     *
     * @param layerName 图层名称
     * @return
     */
    public static LayerSettingGrid getLayerSettingGrid(String layerName) {
        try {
            Layer layer = getLayerByName(layerName);
            //判断是否是专题图
            if (layer != null && layer.getTheme() == null) {
                if (layer.getAdditionalSetting() != null && layer.getAdditionalSetting().getType() == LayerSettingType.GRID) {
                    return (LayerSettingGrid) layer.getAdditionalSetting();
                }
            }
            return null;
        } catch (Exception e) {
            e.printStackTrace();
            return null;
        }
    }

    /**
     * 获取矢量图层设置类
     *
     * @param layerName 图层名称
     * @return
     */
    public static LayerSettingVector getLayerSettingVector(String layerName) {
        try {
            Layer layer = getLayerByName(layerName);
            //判断是否是专题图
            if (layer != null && layer.getTheme() == null) {
                if (layer.getAdditionalSetting() != null && layer.getAdditionalSetting().getType() == LayerSettingType.VECTOR) {
                    return (LayerSettingVector) layer.getAdditionalSetting();
                }
            }
            return null;
        } catch (Exception e) {
            e.printStackTrace();
            return null;
        }
    }


    /**
     * 根据图层名称获取图层对象
     *
     * @return
     */
    public static Layer getLayerByName(String layerName) {
        try {
            MapControl mapControl = SMap.getSMWorkspace().getMapControl();
            Layers layers = mapControl.getMap().getLayers();
            return layers.get(layerName);
        } catch (Exception e) {
            e.printStackTrace();
            return null;
        }
    }

    /**
     * 根据图层索引获取图层对象
     *
     * @return
     */
    public static Layer getLayerByIndex(int layerIndex) {
        try {
            MapControl mapControl = SMap.getSMWorkspace().getMapControl();
            Layers layers = mapControl.getMap().getLayers();
            return layers.get(layerIndex);
        } catch (Exception e) {
            e.printStackTrace();
            return null;
        }
    }
}

File: ios/native/SMCartography.m
//
//  SMCartography.m
//  Supermap
//
//  Created by xianglong li on 2018/12/3.
//  Copyright © 2018 Facebook. All rights reserved.
//

#import "SMCartography.h"
#import "SMap.h"

@implementation SMCartography

+(TextAlignment )getTextAlignment:(NSString*)alignmentText{
    if ([alignmentText isEqualToString:@"TOPLEFT"]) {
        return TA_TOPLEFT;
    }
    if ([alignmentText isEqualToString:@"TOPCENTER"]) {
        return TA_TOPCENTER;
    }
    if ([alignmentText isEqualToString:@"TOPRIGHT"]) {
        return TA_TOPRIGHT;
    }
    if ([alignmentText isEqualToString:@"BASELINELEFT"]) {
        return TA_BASELINELEFT;
    }
    if ([alignmentText isEqualToString:@"BASELINECENTER"]) {
        return TA_BASELINECENTER;
    }
    if ([alignmentText isEqualToString:@"BASELINERIGHT"]) {
        return TA_BASELINERIGHT;
    }
    if ([alignmentText isEqualToString:@"BOTTOMLEFT"]) {
        return TA_BOTTOMLEFT;
    }
    if ([alignmentText isEqualToString:@"BOTTOMCENTER"]) {
        return TA_BOTTOMCENTER;
    }
    if ([alignmentText isEqualToString:@"BOTTOMRIGHT"]) {
        return TA_BOTTOMRIGHT;
    }
    if ([alignmentText isEqualToString:@"MIDDLELEFT"]) {
        return TA_MIDDLELEFT;
    }
    if ([alignmentText isEqualToString:@"MIDDLECENTER"]) {
        return TA_MIDDLECENTER;
    }
    if ([alignmentText isEqualToString:@"MIDDLERIGHT"]) {
        return TA_MIDDLERIGHT;
    }
    return TA_TOPLEFT;
}

+(Geometry *)getGeoText:(Recordset *)recordset{
    @try{
        if (recordset != nil && recordset.recordCount > 0) {
            [recordset moveFirst];
            Geometry* geometry = recordset.geometry;
            if (geometry != nil) {
                if(geometry.getType == GT_GEOTEXT)
                {
                    return geometry;
                }
            }
        }
        return nil;
    }
    @catch(NSException *exception){
        @throw exception;
        return nil;
    }
}
+(Recordset *)getRecordset:(int )geometryID layerName:(NSString *)layerName{
    @try{
        MapControl *mapControl = [SMap singletonInstance].smMapWC.mapControl;
        Layers *layers = mapControl.map.layers;
        Layer *layer = [layers getLayerWithName:layerName];
        if (layer != nil && [layer.dataset isKindOfClass:[DatasetVector class]]) {
            DatasetVector *datasetVector = (DatasetVector *)layer.dataset;
            NSNumber * number = [NSNumber numberWithInt:geometryID];
            NSArray * array = [NSArray arrayWithObject:number];
            return [datasetVector queryWithID:array Type:DYNAMIC];
        }
        return nil;
    }
    @catch(NSException *exception){
        @throw exception;
        return nil;
    }
}
+(LayerSettingGrid *)getLayerSettingGrid:(NSString*)layerName{
    @try{
        Layer *layer = [self getLayerByName:layerName];
        if (layer != nil && layer.theme == nil) {
            if(layer.layerSetting != nil && layer.layerSetting.layerType == GRID)
            {
                layer.editable = true;
                return (LayerSettingGrid* )layer.layerSetting;
            }
        }
        return nil;
    }
    @catch(NSException *exception){
        @throw exception;
        return nil;
    }
}
+(LayerSettingVector *)getLayerSettingVector:(NSString*)layerName{
    @try{
        Layer *layer = [self getLayerByName:layerName];
        if (layer != nil && layer.theme == nil) {
            if(layer.layerSetting != nil && layer.layerSetting.layerType == VECTOR)
            {
                return (LayerSettingVector* )layer.layerSetting;
            }
        }
        return nil;
    }
    @catch(NSException *exception){
        @throw exception;
        return nil;
    }
}

+(LayerSettingVector *)getLayerSettingVectorByIndex:(int)index{
    @try{
        Layer *layer = [self getLayerByIndex:index];
        if (layer != nil && layer.theme == nil) {
            if(layer.layerSetting != nil && layer.layerSetting.layerType == VECTOR)
            {
                return (LayerSettingVector* )layer.layerSetting;
            }
        }
        return nil;
    }
    @catch(NSException *exception){
        @throw exception;
        return nil;
    }
}

+(Layer *)getLayerByName:(NSString*)layerName{
    @try{
        MapControl *mapControl = [SMap singletonInstance].smMapWC.mapControl;
        Layers *layers = mapControl.map.layers;
        return [layers findLayerWithName:layerName];
    }
    @catch(NSException *exception){
        @throw exception;
        return nil;
    }
}

+(Layer *)getLayerByIndex:(int)layerIndex{
    @try{
        MapControl *mapControl = [SMap singletonInstance].smMapWC.mapControl;
        Layers *layers = mapControl.map.layers;
        return [layers getLayerAtIndex:layerIndex];
    }
    @catch(NSException *exception){
        @throw exception;
        return nil;
    }
}

@end
